### PR TITLE
fix(sdk/dataflow): deprecate cluster and use env and platform_instance instead

### PR DESCRIPTION
**Pull request by shubhamjagtap639**
## Deprecation
In Dataflow, cluster argument is deprecated. Please use env instead.

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py [hunk 1]
```diff
@@ -35,11 +52,24 @@ class DataFlow:
     url: Optional[str] = None
     tags: Set[str] = field(default_factory=set)
     owners: Set[str] = field(default_factory=set)
+    platform_instance: Optional[str] = None
+    env: Optional[str] = None
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py:56]: Previously we used `cluster` as the env value - why do we have both cluster and env now? I don't think we need this new instance variable

**Review by hsheth2 — CHANGES_REQUESTED** [review]

File: metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py (hunk 1 above)
Comment (anshbansal) [reply metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py:56]: Plus this is a breaking change and not deprecation. This needs to be documented in the PR doc. Plus we should not be changing the default as part of deprecation as that will cause existing deployments to have problems due to possible change of URNs.

File: metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py [hunk 2]
```diff
@@ -25,6 +25,23 @@
 
 @dataclass
 class DataFlow:
+    """The DataHub representation of data-flow.
+
+    Args:
+        urn (int): Unique identifier of the DataFlow in DataHub. For more detail refer https://datahubproject.io/docs/what/urn/.
+        id (str): Identifier of DataFlow in orchestrator.
+        orchestrator (str): orchestrator. for example airflow.
+        cluster (str): [depricated] Please use env.
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py:34]: typo `depricated`

File: metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py (hunk 1 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py:56]: This isn't a breaking change yet, just a deprecation of `cluster` in favor of `env` + addition of `platform_instance`

However, we still need to make some changes:
1. the deprecation should still be noted in the PR description + our updating datahub doc
2. we should be printing a deprecation warning if folks use the old `cluster` param
3. we should throw an error if both env and cluster are provided
4. all users of this class within datahub should be updated to use `env` - I believe airflow is one, but there might be more

**Review by hsheth2 — CHANGES_REQUESTED** [review]

**Review by mohdsiddique — CHANGES_REQUESTED** [review]

File: docs/how/updating-datahub.md [hunk 3]
```diff
@@ -6,6 +6,8 @@ This file documents any backwards-incompatible changes in DataHub and assists pe
 
 ### Breaking Changes
 
+- #8201: In Dataflow, cluster argument is deprecated. Please use env.
```
Comment (hsheth2) [inline docs/how/updating-datahub.md:9]:
```suggestion
- #8201: Python SDK: In the DataFlow class, the `cluster` argument is deprecated in favor of `env`.
```

File: metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py [hunk 4]
```diff
@@ -45,7 +58,7 @@ class DataFlow:
     urn: DataFlowUrn = field(init=False)
     id: str
     orchestrator: str
-    cluster: str
+    cluster: Optional[str] = None
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py:61]:
```suggestion
    cluster: Optional[str] = None  # Deprecated in favor of env
```

File: metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py [hunk 5]
```diff
@@ -56,21 +69,24 @@ class DataFlow:
...
-                env=self.cluster,
-                flow_id=self.id,
-                platform_instance=self.platform_instance,
+        if self.env is not None and self.cluster is not None:
+            raise ValueError(
+                "Cannot provide env and cluster parameter both at the same time. cluster is deprecated use env."
             )
 
+        if self.env is None and self.cluster is None:
+            raise ValueError("env is required parameter.")
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py:78]:
```suggestion
            raise ValueError("env is required")
```

File: metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py [hunk 6]
```diff
@@ -56,21 +69,24 @@ class DataFlow:
...
-            )
-        else:
-            self.urn = DataFlowUrn.create_from_ids(
-                orchestrator=self.orchestrator,
-                env=self.cluster,
-                flow_id=self.id,
-                platform_instance=self.platform_instance,
+        if self.env is not None and self.cluster is not None:
+            raise ValueError(
+                "Cannot provide env and cluster parameter both at the same time. cluster is deprecated use env."
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py:74]:
```suggestion
                "Cannot provide both env and cluster parameter. Cluster is deprecated in favor of env."
```

File: metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py [hunk 7]
```diff
@@ -56,21 +69,24 @@ class DataFlow:
...
             )
 
+        if self.env is None and self.cluster is None:
+            raise ValueError("env is required parameter.")
+
+        if self.env is None and self.cluster is not None:
+            logger.warning("Use env argument. The cluster argument is deprecated")
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py:81]:
```suggestion
            logger.warning("The cluster argument is deprecated. Use env and possibly platform_instance instead.")
```

Comment (hsheth2) [reply metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py:81]: also just do `self.env = self.cluster`

then you can remove the `if self.env is not None` logic below, and get rid of the casts in `dataprocess_instance.py`

**Review by hsheth2 — CHANGES_REQUESTED** [review]

File: metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py (hunk 7 above)
Comment (shubhamjagtap639) [reply metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py]: self.env is optional, cast is required to pass lint

Comment (shubhamjagtap639) [reply metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py]: self.env is optional, cast is required to pass lint

Comment (shubhamjagtap639) [reply metadata-ingestion/src/datahub/api/entities/datajob/dataflow.py]: self.env is optional, cast is required to pass lint

**Review by hsheth2 — APPROVED** [review]

**Pull request merged**